pandas/core/indexes/period.py

```diff
@@ -835,6 +835,11 @@ def join(self, other, how='left', level=None, return_indexers=False,
         """
         self._assert_can_do_setop(other)
 
+        if not isinstance(other, PeriodIndex):
```

Comment:
what is this needed?

Reply:
hmm, is there a specific test that hits this?

Reply:
this should be in assert_can_do_setop for PeriodIndex

Reply:
`assert_can_do_setop` takes precedence over `_union_incompatible_dtypes`. So if we put it in there, we will no longer be able to union PeriodIndex with anything else, which was one of the original motivators for this PR.

Instead perhaps I can perform the _is_dtype_compatible check here instead.
Or maybe a separate, stricter, `_can_do_join method`?

Reply:
yeah this might need some re-working. a PI can only union with a compatible PI, otherwise its converted to object. But this is true generally for all index types. So not sure why PI needs special logic. Its possible that it is not following the same paths as the other indices (e.g. check DTI & TDI).